File: src/server/api/go/internal/modules/model/task.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type Task struct {
	ID        uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	SessionID uuid.UUID `gorm:"type:uuid;not null;index:ix_session_session_id;index:ix_session_session_id_task_id,priority:1;index:ix_session_session_id_status,priority:1;uniqueIndex:uq_session_id_order,priority:1" json:"session_id"`

	Order         int               `gorm:"not null;uniqueIndex:uq_session_id_order,priority:2" json:"order"`
	Data          datatypes.JSONMap `gorm:"type:jsonb;not null" swaggertype:"object" json:"data"`
	Status        string            `gorm:"type:text;not null;default:'pending';check:status IN ('success','failed','running','pending');index:ix_session_session_id_status,priority:2" json:"status"`
	IsPlanning    bool              `gorm:"not null;default:false" json:"is_planning"`
	SpaceDigested bool              `gorm:"not null;default:false" json:"space_digested"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Task <-> Session
	Session *Session `gorm:"foreignKey:SessionID;references:ID;constraint:OnDelete:CASCADE,OnUpdate:CASCADE;" json:"session"`

	// Task <-> Message (one-to-many)
	Messages []Message `gorm:"constraint:OnDelete:SET NULL,OnUpdate:CASCADE;" json:"messages"`
}

func (Task) TableName() string { return "tasks" }

File: src/server/core/acontext_core/schema/orm/task.py
from dataclasses import dataclass, field
from sqlalchemy import (
    ForeignKey,
    Index,
    Column,
    Integer,
    String,
    Enum,
    text,
    CheckConstraint,
    UniqueConstraint,
    Boolean,
)
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import JSONB, UUID
from typing import TYPE_CHECKING, Optional, List
from .base import ORM_BASE, CommonMixin
from ..session.task import TaskStatus
from ..utils import asUUID

if TYPE_CHECKING:
    from .session import Session
    from .message import Message

# TaskStatusEnum = Enum(TaskStatus, name="task_status_enum", create_type=True)


@ORM_BASE.mapped
@dataclass
class Task(CommonMixin):
    __tablename__ = "tasks"

    __table_args__ = (
        CheckConstraint(
            "status IN ('success', 'failed', 'running', 'pending')",
            name="ck_status",
        ),
        UniqueConstraint(
            "session_id",
            "order",
            name="uq_session_id_order",
        ),
        Index("ix_session_session_id", "session_id"),
        Index("ix_session_session_id_task_id", "session_id", "id"),
        Index("ix_session_session_id_status", "session_id", "status"),
    )

    session_id: asUUID = field(
        metadata={
            "db": Column(
                UUID(as_uuid=True),
                ForeignKey("sessions.id", ondelete="CASCADE"),
                nullable=False,
            )
        }
    )

    order: int = field(metadata={"db": Column(Integer, nullable=False)})

    data: dict = field(metadata={"db": Column(JSONB, nullable=False)})

    status: str = field(
        default="pending",
        metadata={
            "db": Column(
                String,
                nullable=False,
                server_default="pending",
            )
        },
    )

    is_planning: bool = field(
        default=False,
        metadata={
            "db": Column(Boolean, nullable=False, default=False, server_default="false")
        },
    )

    space_digested: bool = field(
        default=False,
        metadata={
            "db": Column(Boolean, nullable=False, default=False, server_default="false")
        },
    )

    # Relationships
    messages: List["Message"] = field(
        default_factory=list,
        metadata={"db": relationship("Message", back_populates="task")},
    )

    session: "Session" = field(
        init=False,
        metadata={"db": relationship("Session", back_populates="tasks")},
    )
